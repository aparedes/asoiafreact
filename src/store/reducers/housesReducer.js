/* @flow */

import * as Immutable from 'immutable'
import type { List, Map, Set } from 'immutable'

type ActionType = {
  type: string,
  allHouses: Map<string, *>,
  allHousesIds: List<string>,
  houseId: string,
  house: Map<string, *>,
  error: string,
  regions: Set<string>,
}
export default function (state: Map<string, *> = new Immutable.Map(), action: ActionType) {
  if (action) {
    switch (action.type) {
      // ALL HOUSES
      case 'GET_ALL_HOUSES':
        return state.set('getting_all', true)
      case 'GET_ALL_HOUSES_ERROR':
        return state.remove('getting_all')
                    .set('error_all', action.error)
      case 'GOT_ALL_HOUSES':
        return state.remove('getting_all')
                    .set('houses', action.allHouses)
                    .set('housesIds', action.allHousesIds)
                    .set('regions', action.regions)
      // SINGLE HOUSE
      case 'GET_HOUSE':
        return state.setIn([ 'houses', action.houseId, 'getting' ], true)
      case 'GET_HOUSE_ERROR':
        return state.removeIn([ 'houses', action.houseId, 'getting' ])
                    .setIn([ 'houses', action.houseId, 'error' ], action.error)
      case 'GOT_HOUSE':
        return state.mergeIn([ 'houses', action.houseId ], action.house)
                    .removeIn([ 'houses', action.houseId, 'getting' ])
                    .removeIn([ 'houses', action.houseId, 'error' ])
      default:
    }
  }
  return state
}
