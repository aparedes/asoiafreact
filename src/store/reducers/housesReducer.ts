/* @flow */

import { createSlice, PayloadAction } from '@reduxjs/toolkit';

type House = {
  id: string;
  region: string;
  name: string;
  currentLord?: {
    name: string;
  };
  coatOfArms?: string;
  words?: string;
  error?: string;
  getting?: boolean;
};

export type HouseState = {
  gettingAll: boolean;
  errorAll?: string;
  housesIds?: string[];
  houses: House[];
  regions?: Readonly<string[]>;
};

const { actions, reducer } = createSlice({
  name: 'house',
  initialState: {
    gettingAll: false,
    houses: [],
    regions: [],
    housesIds: [],
  } as HouseState,
  reducers: {
    getAllHouses(state) {
      state.gettingAll = true;
    },
    getAllHousesError(state, action) {
      state.gettingAll = false;
      state.errorAll = action.payload;
    },
    gotAllHouses(
      state,
      action: PayloadAction<{
        allHouses: House[];
        allHousesIds: string[];
        regions: string[];
      }>
    ) {
      state.gettingAll = false;
      state.houses = action.payload.allHouses;
      state.housesIds = action.payload.allHousesIds;
      state.regions = action.payload.regions;
    },
    getHouse(state, action: PayloadAction<string>) {
      const houseId = state.houses.findIndex(
        (house) => house.id === action.payload
      );
      ((state.houses[houseId] as Record<string, unknown>) ?? {}).getting = true;
    },
    getHouseError(
      state,
      action: PayloadAction<{ houseId: string; error: string }>
    ) {
      const houseId = state.houses.findIndex(
        (house) => house.id === action.payload.houseId
      );
      ((state.houses[houseId] as Record<string, unknown>) ?? {}).getting =
        false;
      ((state.houses[houseId] as Record<string, unknown>) ?? {}).error =
        action.payload.error;
    },
    gotHouse(
      state,
      action: PayloadAction<{ houseId: string; house: Record<string, unknown> }>
    ) {
      const houseId = state.houses.findIndex(
        (house) => house.id === action.payload.houseId
      );
      ((state.houses[houseId] as Record<string, unknown>) ?? {}).getting =
        false;
      ((state.houses[houseId] as Record<string, unknown>) ?? {}).error = null;
      (state.houses[houseId] as Record<string, unknown>) = {
        ...state.houses[houseId],
        ...action.payload.house,
      };
    },
  },
});

export const houseReducer = reducer;
export const houseActions = actions;
