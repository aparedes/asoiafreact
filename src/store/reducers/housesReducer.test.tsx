
import { houseActions, houseReducer } from './housesReducer';

describe('Test House Reducer', () => {
  const map = { gettingAll: false, houses: [] };
  // it('Returns Map from unknown action', () => {
  //   let action = { type: 'INIT_REDUX' };
  //   expectIs(reducer(map, action), Map());
  // });

  it('GET_ALL_HOUSES', () => {
    expect(houseReducer(undefined, houseActions.getAllHouses())).toEqual({
      gettingAll: true,
      houses: [],
      housesIds: [],
      regions: [],
    });
  });

  it('GET_ALL_HOUSES_ERROR after GET_ALL_HOUSES', () => {
    expect(
      houseReducer(
        { gettingAll: true, houses: [] },
        houseActions.getAllHousesError('Fetch Error')
      )
    ).toEqual({
      gettingAll: false,
      houses: [],
      errorAll: 'Fetch Error',
    });
  });

  it('GET_ALL_HOUSES after GET_ALL_HOUSES', () => {
    let action = {
      allHouses: [
        { id: '1', name: 'House Stark of Winterfell', region: 'The North' },
      ],
      allHousesIds: ['1'],
      regions: ['The North'],
    };
    expect(
      houseReducer(map, houseActions.gotAllHouses(action))
    ).toMatchSnapshot();
  });
});
