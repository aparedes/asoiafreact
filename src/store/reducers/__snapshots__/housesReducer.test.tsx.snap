// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test House Reducer GET_ALL_HOUSES after GET_ALL_HOUSES 1`] = `
Object {
  "gettingAll": false,
  "houses": Array [
    Object {
      "id": "1",
      "name": "House Stark of Winterfell",
      "region": "The North",
    },
  ],
  "housesIds": Array [
    "1",
  ],
  "regions": Array [
    "The North",
  ],
}
`;
